File: 1b/statistics_long.pl
% count_occurrences/3 counts the number of occurrences of an element in a list
count_occurrences(_, [], 0).
count_occurrences(X, [X|T], N) :-
    count_occurrences(X, T, N1),
    N is N1 + 1.
count_occurrences(X, [Y|T], N) :-
    X \= Y,
    count_occurrences(X, T, N).

% array_statistics/2 calculates the appearance statistics of each element in a list
array_statistics(Array, Stats) :-
    remove_duplicates(Array, Unique),
    array_statistics_helper(Unique, Array, Stats).

% Helper predicate for array_statistics/2
array_statistics_helper([], _, []).
array_statistics_helper([X|Unique], Array, [(X, Count)|Stats]) :-
    count_occurrences(X, Array, Count),
    array_statistics_helper(Unique, Array, Stats).

% Remove duplicates from a list
remove_duplicates([], []).
remove_duplicates([X|Xs], Unique) :-
    member(X, Xs),
    remove_duplicates(Xs, Unique).
remove_duplicates([X|Xs], [X|Unique]) :-
    \+ member(X, Xs),
    remove_duplicates(Xs, Unique).

File: 1b/statistics_long.hs
-- count_occurrences counts the number of occurrences of an element in a list
count_occurrences :: Eq a => a -> [a] -> Int
count_occurrences _ [] = 0
count_occurrences x (y:ys)
    | x == y = 1 + count_occurrences x ys
    | otherwise = count_occurrences x ys

-- array_statistics calculates the appearance statistics of each element in a list
array_statistics :: Eq a => [a] -> [(a, Int)]
array_statistics [] = []
array_statistics (x:xs) =
    let occurrences = count_occurrences x (x:xs)
        rest_stats = array_statistics (removeFirstOccurrence x xs)
    in (x, occurrences) : rest_stats

-- Remove the first occurrence of an element from the list
removeFirstOccurrence :: Eq a => a -> [a] -> [a]
removeFirstOccurrence _ [] = []
removeFirstOccurrence x (y:ys)
    | x == y = ys
    | otherwise = y : removeFirstOccurrence x ys


array4 :: [Int]
array4 = []

main :: IO ()
main = do
    let array = [1, 2, 3, 2, 1, 4, 5, 3, 2]
        stats = array_statistics array
    putStrLn $ "Array: " ++ show array
    putStrLn $ "Statistics: " ++ show stats

    let array2 = [1]
        stats2 = array_statistics array2
    putStrLn $ "Array: " ++ show array2
    putStrLn $ "Statistics: " ++ show stats2

    let array3 = [1, 2, 1, 2]
        stats3 = array_statistics array3
    putStrLn $ "Array: " ++ show array3
    putStrLn $ "Statistics: " ++ show stats3

    let stats4 = array_statistics array4
    putStrLn $ "Array: " ++ show array4
    putStrLn $ "Statistics: " ++ show stats4
